README.md
# f4clojure

attempting problems from 4clojure.com

`lein test` runs through the tests for all the solutions


CHANGELOG.md
Attempting problems from 4clojure.com


test/f4clojure/core_test.clj
(ns f4clojure.core-test
  (:require [clojure.test :refer :all]
            [f4clojure.core :refer :all]))

(deftest test-last-element-in-sequence
  (testing "simple vector"
    (is (= 5 (last-element-in-sequence [1 2 3 4 5]))))
  (testing "simple list"
    (is (= 3 (last-element-in-sequence '(5 4 3)))))
  (testing "strings in vector"
    (is (= "d" (last-element-in-sequence ["b" "c" "d"]))))
  (testing "empty vector"
    (is (= nil (last-element-in-sequence [])))))

(deftest test-penultimate-element
  (testing "simple vector"
    (are [x y] (= x y)
      10 (penultimate [1 2 3 10 100])
      :a (penultimate [:a :b])
      nil (penultimate [1]))))

(deftest test-nth-elem
  (are [x y] (= x y)
    1 (nth-elem [9 1 8 2] 1)
    1 (nth-elem [1 2 3 4] 0)
    nil (nth-elem [] 1)
    nil (nth-elem [1 2 3] 1000)))

(deftest test-my-count
  (are [x y] (= x y)
    1 (my-count [1])
    0 (my-count [])
    4 (my-count [1 2 2 2])))


(deftest test-reverse-seq
  (are [x y] (= x y)
    '(3 2 1) (reverse-seq [1 2 3])
    '(1) (reverse-seq [1])
    '() (reverse-seq [])))

(deftest test-sum-up
  (are [x y] (= x y)
    0 (sum-up [1 -1 2 -2])
    10 (sum-up [1 2 3 4])))

(deftest test-flip-order
  (are [x y] (= x y)
    false (> 7 8)
    true ((flip-order >) 7 8)
    "hello world" (str "hello" " " "world")
    "world hello" ((flip-order str) "hello" " " "world")))

(deftest test-rotate-seq
  (let [s [1 2 3 4 5]]
    (are [x y] (= x y)
      [3 4 5 1 2] (rotation s 2)
      s (rotation s 5)
      [4 5 1 2 3] (rotation s -2))))

(deftest test-reverse-interleave
  (are [x y] (= x y)
    '([1 4 7] [2 5 8] [3 6 9]) (reverse-interleave (range 1 10) 3)
    '((0 5) (1 6) (2 7) (3 8) (4 9)) (reverse-interleave (range 10) 5)))

(deftest test-type-split
  (are [x y] (= x y)
    '([1 2 3] [:a :b :c]) (split-by-type [1 :a 2 :b 3 :c])
    '([:a :b] ["foo" "bar"]) (split-by-type [:a "foo"  "bar" :b])
    '([[1 2] [3 4]] [:a :b] [5 6]) (split-by-type [[1 2] :a [3 4] 5 6 :b])))

(deftest test-count-items
  (are [x y] (= x y)
    {1 1 2 2 3 3 4 4} (count-items [1 2 2 3 3 3 4 4 4 4])
    {} (count-items [])))

(deftest test-remove-dups
  (are [x y] (= x y)
    [1 2] (remove-dups [1 2 1 2 1 2 1 1 1 1 1 2 2 2 2 2])
    [1 2 3] (remove-dups [1 2 2 2 1 1 1 1 3 3 3 3 2 2 1 1])
    (range 50) (remove-dups (range 50))))

(deftest test-compose-funcs
  (are [x y] (= x y)
    5 ((composer count last) ["aaaa" "bbbbb" "ccccc"])
    100 ((composer (partial + 99) first) [1 :a :x])
    false ((composer empty? first reverse) [[] [2] [444]])))

(deftest test-part-seq
  (are [x y] (= x y)
    [[0 1] [2 3] [4 5] [6 7] [8 9]] (part-seq 2 (range 10))
    [[0 1 2 3] [4 5 6 7]] (part-seq 4 (range 10))))

(deftest test-juxtapose
  (are [x y] (= x y)
    [21 6 1] ((juxtapose + max min) 2 3 5 1 6 4)
    ["HELLO" 5] ((juxtapose #(.toUpperCase %) count) "hello")
    [2 6 4] ((juxtapose :a :c :b) {:a 2, :b 4, :c 6, :d 8 :e 10})))

(deftest test-split-and-sort
  (are [x y] (= x y)
    ["a" "conundrum" "is" "quite" "This"] (split-sentence-and-sort "This is quite, a conundrum!")
    ["fall" "follies" "foolish" "Fools" "for"] (split-sentence-and-sort  "Fools fall for foolish follies.")))

(deftest test-n-primes
  (are [x y] (= x y)
    [2 3 5 7] (n-primes 10)
    [2 3 5 7 11 13 17 19] (n-primes 20)))

(deftest test-black-box
  (are [x y] (= x y)
    :set (black-box-testing #{1 :a})
    :map (black-box-testing {:a 1})
    :vec (black-box-testing [])
    :list (black-box-testing '(1 2 3))))

(deftest test-filter-squares
  (are [x y] (= x y)
    "16,144" (filter-squares "12,16,45,144,991")
    "" (filter-squares "12,13,14")))

(deftest test-trampoline
  (are [x y] (= x y)
    [1 3 5 7 9 11]
    (letfn
        [(foo
           [x y]
           #(bar (conj x y) y))
         (bar
           [x y]
           (if (> (last x) 10)
             x
             #(foo x (+ 2 y))))]
      (my-trampoline foo [] 1))))

(deftest test-anagram-finder
  (are [x y] (= x y)
    #{#{"meat" "team" "mate"}} (anagram-finder ["meat" "mat" "team" "mate" "eat"])
    #{#{"veer" "ever"} #{"lake" "kale"} #{"mite" "item"}} (anagram-finder ["veer" "lake" "item" "kale" "mite" "ever"])))

(deftest test-seq-reduce
  (are [x y] (= x y)
    [0 1 3 6 10] (take 5 (seq-reduce + (range)))
    [[1] [1 2] [1 2 3] [1 2 3 4]] (seq-reduce conj [1] [2 3 4])
    (reduce * 2 [3 4 5]) (last (seq-reduce * 2 [3 4 5]))))

(deftest test-perfect
  (are [x y] (= x y)
    true (perfect? 6)
    true (perfect? 496)
    false (perfect? 500)))

(deftest test-merge-maps
  (are [x y] (= x y)
    {:a 4, :b 6, :c 20} (merge-maps * {:a 2, :b 3, :c 4} {:a 2} {:b 2} {:c 5})
    {1 7, 2 10, 3 15} (merge-maps - {1 10, 2 20} {1 3, 2 10, 3 15})))

(deftest test-camel-case
  (are [x y] (= x y)
    "something" (to-camel-case "something")
    "multiWordKey" (to-camel-case "multi-word-key")
    "leaveMeAlone" (to-camel-case "leaveMeAlone")))

(deftest test-totient
  (are [x y] (= x y)
    16 (euler-totient 40)
    60 (euler-totient 99)
    4 (euler-totient 10)))

(deftest test-happy
  (are [x y] (= x y)
    true (happy? 7)
    false (happy? 2)
    true (happy? 986543210)))

(deftest test-re-trampoline
  (are [x y] (= x y)
    82 (letfn [(triple [x] #(sub-two (* 3 x)))
            (sub-two [x] #(stop?(- x 2)))
            (stop? [x] (if (> x 50) x #(triple x)))]
         (re-trampoline triple 2))
    [true false true false true false]
    (letfn [(my-even? [x] (if (zero? x) true #(my-odd? (dec x))))
            (my-odd? [x] (if (zero? x) false #(my-even? (dec x))))]
      (map (partial re-trampoline my-even?) (range 6)))))

(deftest test-balanced
  (are [x y] (= x y)
    true (balanced? 11)
    true (balanced? 121)
    false (balanced? 123)
    true (balanced? 0)
    false (balanced? 88099)
    true (balanced? 89098)
    true (balanced? 89089)
    [0 1 2 3 4 5 6 7 8 9 11 22 33 44 55 66 77 88 99 101]
    (take 20 (filter balanced? (range)))))

(deftest test-powerset
  (are [x y] (= x y)
    #{#{1 :a} #{:a} #{} #{1}} (power-set #{1 :a})
    #{#{}} (power-set #{})
    #{#{} #{1} #{2} #{3} #{1 2} #{1 3} #{2 3} #{1 2 3}} (power-set #{1 2 3})
    1024 (count (power-set (into #{} (range 10))))))

(deftest test-equivalence
  (are [x y] (= x y)
    #{#{0} #{1 -1} #{2 -2}} (equivalence #(* % %) #{-2 -1 0 1 2})
    #{#{0 3} #{1 4} #{2 5}} (equivalence #(rem % 3) #{0 1 2 3 4 5 })
    #{#{0} #{1} #{2} #{3} #{4}} (equivalence identity #{0 1 2 3 4})
    #{#{0 1 2 3 4}} (equivalence (constantly true) #{0 1 2 3 4})))

(deftest test-map-with-vals
  (are [x y] (= x y)
    {:a [1 2 3], :b [], :c [4]} (map-with-vals [:a 1 2 3 :b :c 4])
    {:a [1], :b [2]} (map-with-vals [:a 1, :b 2])
    {:a [1]} (map-with-vals [:a 1])))

(deftest test-seq-base
  (are [x y] (= x y)
    [1 2 3 4 5 0 1] (seq-base 1234501 10)
    [0] (seq-base 0 11)
    [1 0 0 1] (seq-base 9 2)
    [1 0] (let [n (rand-int 100000)](seq-base n n))
    [16 18 5 24 15 1] (seq-base Integer/MAX_VALUE 42)))


misc/init.el
(package-initialize)

(setq package-archives
      '(("gnu" . "http://elpa.gnu.org/packages/")
	("melpa" . "http://melpa.org/packages/")))


(show-paren-mode 1)
(global-company-mode)
(global-linum-mode)
(ido-mode t)

(add-hook 'cider-repl-mode-hook #'cider-company-enable-fuzzy-completion)

(add-hook 'cider-mode-hook #'cider-company-enable-fuzzy-completion)
(add-hook 'cider-mode-hook 'eldoc-mode)
(add-hook 'cider-mode-hook 'paredit-mode)
(add-hook 'cider-mode-hook 'rainbow-delimiters-mode)

(add-hook 'clojure-mode-hook #'cider-company-enable-fuzzy-completion)
(add-hook 'clojure-mode-hook 'eldoc-mode)
(add-hook 'clojure-mode-hook 'paredit-mode)
(add-hook 'clojure-mode-hook 'rainbow-delimiters-mode)

(setq company-idle-delay 0.1)
(setq company-minimum-prefix-length 2)

(setq make-backup-files nil)
(setq auto-save-default nil)

(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(package-selected-packages (quote (magit company cider paredit rainbow-delimiters))))


src/f4clojure/core.clj
(ns f4clojure.core
  (:require [clojure.string :as s])
  (:gen-class))

;; http://www.4clojure.com/problem/19
(defn last-element-in-sequence
  [sequence] 
  (reduce #(identity %2) nil sequence))

;; http://www.4clojure.com/problem/20
(defn penultimate
  [elems]
  (nth (reverse elems) 1 nil))

;; http://www.4clojure.com/problem/12
(defn nth-elem
  [elems n]
  (if (< (count elems) n)
    nil
    (last (take (inc n) elems))))

;; http://www.4clojure.com/problem/22
(defn my-count
  [elems]
  (reduce (fn [a b] (inc a)) 0 elems))

;; http://www.4clojure.com/problem/23
(defn reverse-seq
  [s]
  (reduce #(cons %2 %1) [] s))

;; http://www.4clojure.com/problem/24
(defn sum-up [s] (apply + s))

;; http://www.4clojure.com/problem/46
(defn flip-order
  [f]
  (fn [& args]
    (apply f (reverse args))))

(defn dispatch-rotate
  [_ times]
  (cond
    (> times 0) :counter-clock-wise
    (< times 0) :clock-wise))

;; http://www.4clojure.com/problem/44
(defmulti rotation dispatch-rotate)

(defmethod rotation :counter-clock-wise
  [s times]
  (flatten (conj
            (vec (drop times s))
            (take times s))))

(defmethod rotation :clock-wise
  [s times]
  (let [times (- times)]
    (flatten (cons
              (take-last times s)
              (drop-last times s)))))

;; http://www.4clojure.com/problem/43
(defn reverse-interleave
  "First we create a list of sub-lists, such that each sub-list starts at
  further and further indices. Then we select those sublists that can give
  us the right items (the first chunk-size).

  Finally we call take-nth on each of these sub-lists to give the result.
  "
  [items chunk-size]
  (let [items-size (count items)
        sub-lists (map #(drop % items) (range items-size))
        sub-lists-large-enough (take chunk-size sub-lists)]
    (map #(take-nth chunk-size %) sub-lists-large-enough)))

;; http://www.4clojure.com/problem/50
(defn split-by-type
  [items]
  (vals (group-by type items)))

;; http://www.4clojure.com/problem/55
(defn count-items
  [items]
  (->>
   items
   (group-by identity)
   (#(zipmap (keys %) (map count (vals %))))))

;; http://www.4clojure.com/problem/56
(defn remove-dups
  [items]
  (reduce
   (fn [uniques item]
     (if (some #(= item %) uniques)
       uniques
       (conj uniques item)))
   []
   items))

;; http://www.4clojure.com/problem/58
(defn composer
  [& funcs]
  (let [funcs (reverse funcs)]
    (reduce (fn [composed f]
              (fn [& args] (f (apply composed args))))
            (first funcs)
            (rest funcs))))

;; http://www.4clojure.com/problem/54
(defn part-seq
  [size items]
  (loop [items items
         chunks []]
    (if (< (count items) size)
      chunks
      (let [[a b] (split-at size items)]
        (recur b (conj chunks a))))))

;; http://www.4clojure.com/problem/59
(defn juxtapose
  [& functions]
  (fn [& args]
    (map #(apply % args) functions)))

;; http://www.4clojure.com/problem/70
(defn split-sentence-and-sort
  [sentence]
  (sort-by #(.toLowerCase %) (re-seq #"[a-zA-Z0-9]+" sentence))) ; not sure if using regexes is allowed in problems


(defn prime?
  [num]
  (let [limit (inc (quot num 2))]
  (nil?
   (some #(= (rem num %) 0)
         (range 2 limit)))))

;; http://www.4clojure.com/problem/67
(defn n-primes
  [n]
  (loop [n n current 2 primes []]
    (if (= 2 n) ; because we start at 2, we should end early
      primes
      (recur (dec n) (inc current)
             (if (prime? current) (conj primes current) primes)))))

;; http://www.4clojure.com/problem/65
(defn black-box-testing
  [arg]
  (letfn [(is-list? [arg]
            (let [a (rand-int 10000) b (rand-int 10000) c (conj arg a b)]
              (= (first c) b)))
          (is-vec? [arg]
            (let [a (rand-int 10000) b (rand-int 10000) c (conj arg a b)]
              (= (last c) b)))
          (is-set? [arg]
            (let [f (first arg) size (count arg)]
              (= (count (conj arg f)) size)))
          (is-map? [arg]
            (try
              (conj arg 0)
              false ; if conj worked, its not a map
              (catch IllegalArgumentException e true)))]
    (cond
      (is-map? arg) :map
      (is-set? arg) :set
      (is-vec? arg) :vec
      (is-list? arg) :list)))

;; http://www.4clojure.com/problem/74
(defn filter-squares
  [s]
  (letfn [(square? [n] (some #(= (* % %) n) (range (quot n 2))))]
    (->>
     s
     (#(s/split % #","))
     (map #(Integer/parseInt %))
     (filter square?)
     (interpose ",")
     (apply str))))

;; http://www.4clojure.com/problem/76
(defn my-trampoline
  [f & args]
  (let [a (apply f args)]
    (some
     #(if (not (fn? %)) %)
     (iterate #(if (fn? %) (%) %) a))))

;; http://www.4clojure.com/problem/77
(defn anagram-finder
  [words]
  (set (->>
        words
        (group-by set)
        vals
        (map set)
        (filter #(> (count %) 1)))))

;; http://www.4clojure.com/problem/60
(defn seq-reduce
  "Only got this to work with lazy sequences after looking at code for reductions function"
  ([f args]
   (seq-reduce f (first args) (rest args)))
  ([f elem args]
   (lazy-seq (cons elem
                   (when-let [s (seq args)]
                     (seq-reduce f
                                 (f elem (first s))
                                 (rest s)))))))

(defn sum-of-divisors [n]
  (let [bounds (range 1 (inc (quot n 2)))]
    (apply + (reduce (fn [divisors x] (if (= 0 (rem n x)) (conj divisors x) divisors))
                     #{}
                     bounds))))

;; http://www.4clojure.com/problem/80
(defn perfect? [n]
  (= (sum-of-divisors n) n))

;; http://www.4clojure.com/problem/69
(defn merge-maps [f & maps]
  (reduce
   (fn [merged-map m]
     (let [common-keys (clojure.set/intersection
                        (set (keys merged-map))
                        (set (keys m)))
           pre-merge (conj merged-map m)]	;; start with a naive merge of all keys
       (reduce
        (fn [final-map key]			;; for all common keys rewrite with call to f
          (assoc final-map key (f (merged-map key) (m key))))
        pre-merge
        common-keys)))
   {}
   maps))

;; http://www.4clojure.com/problem/102
(defn to-camel-case [s]
  (let [words (->>
               s
               (partition-by #(= \- %))
               (filter #(not= '(\-) %))
               (map #(apply str %)))]
    (str (first words)
         (apply str (map s/capitalize (rest words))))))

(defn gcd [a b]
  (if (= 0 a)
    b
    (recur (mod b a) a)))

(defn co-prime? [a b]
  (= 1 (gcd a b)))

;; http://www.4clojure.com/problem/75
(defn euler-totient [n]
  (if (= 1 n)
    1
    (count (filter #(co-prime? n %) (range 1 n)))))

;; http://www.4clojure.com/problem/86
(defn happy? [n]
  (letfn [(split-num [n]
            (->>
             n
             (iterate #(quot % 10))
             (take-while pos?)
             (map #(mod % 10))
             reverse))]
    (loop [n n cache #{}]
      (cond
        (= n 1) true
        (contains? cache n) false
        :else (recur (apply +
                      (map #(* % %)
                           (split-num n)))
                     (conj cache n))))))

;; http://www.4clojure.com/problem/78
(defn re-trampoline [f & args]
  (if (not (fn? f))
    f
    (recur (apply f args) nil)))

;; http://www.4clojure.com/problem/115
(defn balanced? [n]
  (let [pieces (->>
                (iterate #(quot % 10) n)
                (take-while pos?)
                (map #(mod % 10))
                reverse)
        mid (quot (count pieces) 2)
        [top bottom] (split-at mid pieces)
        bottom (if (= (count top)
                      (count bottom))
                 bottom
                 (rest bottom))]
    (if (= mid 0)
      true
      (= (apply + top) (apply + bottom)))))

;; http://www.4clojure.com/problem/85
(defn power-set [s]
  "Power set algorithm from wikipedia"
  (if (empty? s)
    #{s}
    (let [f (first s)
          r (disj s f)
          rs (power-set r)]
      (apply conj rs (map #(conj % f) rs)))))

;; http://www.4clojure.com/problem/98
(defn equivalence [f s]
  (set (map set (vals (group-by f s)))))

;; http://www.4clojure.com/problem/105
(defn map-with-vals [coll]
  (reduce (fn [m e]
            (if (keyword? e)
              (assoc m e [])
              (let [[k v] (last m)
                    new-v (conj v e)]
                (assoc m k new-v))))
          (array-map)
          coll))

;; http://www.4clojure.com/problem/137
(defn seq-base [n base]
  (loop [nums (list (rem n base))
         n (quot n base)]
    (if (pos? n)
      (recur (cons (rem n base) nums)
             (quot n base))
      nums)))
